kornia/feature/scale_space_detector.py

```diff
@@ -46,7 +45,7 @@ def _create_octave_mask(mask: torch.Tensor, octave_shape: List[int]) -> torch.Te
     return mask_octave.unsqueeze(1)
 
 
-class ScaleSpaceDetector(nn.Module):
+class ScaleSpaceDetector(kornia.nn.ScaleSpaceDetector):
```

Comment:
@shijianjian moving all the `nn.Module` functionalities like this one I think needs proper discussion since the reason of being a module is that they hold a state or e.g in the case for filters they have a stateful kernel. Doing this refactor will break I believe a backward compatibility and needs proper thinking and discussion. First, we should define what's is the goal for `kornia.nn` modules  since for now it's just wrappers around functionals and I believe what `kornia.nn` should be it's more than that, and we even could provide both: object wrappers and pytorch-like modules with "auto" learnable parameters. I would like discuss carefully about this before merging this PR. Big PR moving things around can lead to bigger maintenance issues later if are not well defined and discussed. /cc @ducha-aiki @anguelos @pmeier

Reply:
This PR does nothing but moving things to ```kornia.nn``` and deprecating the current. I think this is required for any further decisions as long as there is a ```kornia.nn``` module.

For refactoring the filters module, I do not think it will break anything. For example, the boxblur:
```python
class BoxBlur(nn.Module):
    def __init__(self, kernel_size: Tuple[int, int],
                 border_type: str = 'reflect',
                 normalized: bool = True) -> None:
        super(BoxBlur, self).__init__()
        self.kernel_size: Tuple[int, int] = kernel_size
        self.border_type: str = border_type
        self.kernel: torch.Tensor = get_box_kernel2d(kernel_size)
        self.normalized: bool = normalized
        if self.normalized:
            self.kernel = normalize_kernel2d(self.kernel)

    def __repr__(self) -> str:
        return self.__class__.__name__ +\
            '(kernel_size=' + str(self.kernel_size) + ', ' +\
            'normalized=' + str(self.normalized) + ', ' + \
            'border_type=' + self.border_type + ')'

    def forward(self, input: torch.Tensor):  # type: ignore
        return kornia.filter2D(input, self.kernel, self.border_type)


def box_blur(input: torch.Tensor,
             kernel_size: Tuple[int, int],
             border_type: str = 'reflect',
             normalized: bool = True) -> torch.Tensor:
    return BoxBlur(kernel_size, border_type, normalized)(input)
```

This is clearly an improper implementation which would create a ```nn.Module``` for a functional call. A refactoring could be:

```python
class BoxBlur(nn.Module):
    def __init__(self, kernel_size: Tuple[int, int],
                 border_type: str = 'reflect',
                 normalized: bool = True) -> None:
        super(BoxBlur, self).__init__()
        self.kernel_size: Tuple[int, int] = kernel_size
        self.border_type: str = border_type
        self.normalized: bool = normalized

    def __repr__(self) -> str:
        return self.__class__.__name__ +\
            '(kernel_size=' + str(self.kernel_size) + ', ' +\
            'normalized=' + str(self.normalized) + ', ' + \
            'border_type=' + self.border_type + ')'

    def forward(self, input: torch.Tensor):  # type: ignore
        return box_blur(input, self.kernel_size, self.border_type, self.normalized)


def box_blur(input: torch.Tensor,
             kernel_size: Tuple[int, int],
             border_type: str = 'reflect',
             normalized: bool = True) -> torch.Tensor:
     kernel: torch.Tensor = get_box_kernel2d(kernel_size)
     if normalized:
        kernel = normalize_kernel2d(self.kernel)
    return kornia.filter2D(input, kernel, border_type)
```

For the goal of ```kornia.nn```, I think it will have features like:
1. to be used in ```nn.Sequential```.
2. exporting ONNX model.
3. param learning for needed module.
4. (optional to me) exporting JIT.

I do not think it is necessary to provide the object-wrappers since the concept of kornia lib aims at differentiable computer vision.